nautobot_ssot/metrics.py

```diff
@@ -0,0 +1,156 @@
+"""Nautobot SSoT framework level metrics."""
+from django.conf import settings
+from prometheus_client.core import GaugeMetricFamily
+from nautobot.extras.models.jobs import Job
+from nautobot_ssot.models import Sync
+
+PLUGIN_SETTINGS = settings.PLUGINS_CONFIG.get("nautobot_ssot", {})
+
+
+def metric_ssot_jobs():
+    """Extracts duration of latest SSoT Job run.
+
+    Yields:
+        GaugeMetricFamily: Prometheus Metrics
+    """
+    ssot_job_durations = GaugeMetricFamily(
+        "nautobot_ssot_duration_microseconds",
+        "Nautobot SSoT Job Phase Duration in microseconds",
+        labels=["phase", "job"],
+    )
+
+    for job in Job.objects.filter(slug__icontains="ssot"):
+        last_job_sync = Sync.objects.filter(job_result__job_model_id=job.id).last()
+        if last_job_sync:
+            if last_job_sync.source_load_time:
+                ssot_job_durations.add_metric(
+                    labels=["source_load_time", job.slug],
+                    value=(last_job_sync.source_load_time.seconds * 100000)
+                    + last_job_sync.source_load_time.microseconds,
+                )
+
+            if last_job_sync.target_load_time:
+                ssot_job_durations.add_metric(
+                    labels=["target_load_time", job.slug],
+                    value=(last_job_sync.target_load_time.seconds * 1000000)
+                    + last_job_sync.target_load_time.microseconds,
+                )
+
+            if last_job_sync.diff_time:
+                ssot_job_durations.add_metric(
+                    labels=["diff_time", job.slug],
+                    value=(last_job_sync.diff_time.seconds * 1000000) + last_job_sync.diff_time.microseconds,
+                )
+
+            if last_job_sync.sync_time:
+                ssot_job_durations.add_metric(
+                    labels=["sync_time", job.slug],
+                    value=(last_job_sync.sync_time.seconds * 1000000) + last_job_sync.sync_time.microseconds,
+                )
+
+            if last_job_sync.duration:
+                ssot_job_durations.add_metric(
+                    labels=["sync_duration", job.slug],
+                    value=(last_job_sync.duration.seconds * 1000000) + last_job_sync.duration.microseconds,
+                )
+
+    yield ssot_job_durations
+
+
+def metric_syncs():
+    """Calculate total number of Syncs.
+
+    Yields:
+        GaugeMetricFamily: Prometheus Metrics
+    """
+    sync_gauge = GaugeMetricFamily("nautobot_ssot_sync_totals", "Nautobot SSoT Sync Totals", labels=["sync_type"])
+
+    sync_gauge.add_metric(labels=["total_syncs"], value=Sync.objects.all().count())
+
+    for status_type in ("completed", "failed", "errored", "pending", "running"):
```

Comment:
Should this be iterating over a choices variable from core?

Reply:
This isn't a choice option that a user can select. It's automatically set from the Job results.

Reply:
My suggestion was to loop over this maybe: https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/choices.py#L189

Then if something changes in Nautobot core the metric is automatically appended to.

Reply:
Done.